### ISSUE TITLE
Many i386 builds in "previous builds"

### ISSUE THREAD

**Issue by evandandrea** [canonical-websites/build.snapcraft.io#983]
![screen shot 2017-10-25 at 11 54 13](https://user-images.githubusercontent.com/1523179/31995200-5f1aad4c-b97b-11e7-8deb-f973282bfdc9.png)

This seems odd to me. Is there a reasonable explanation for why we're seeing one amd64, one armhf, and five i386 builds in the "previous builds" section?

**Comment by cjwatson** [comment]
Yes, there is.  There've been several commits recently, each of which scheduled a build for each architecture.  But we had outages in both our x86 scalingstack regions between me finishing work yesterday and starting work this morning, so all amd64 and i386 builds ended up being backlogged for a while.  We restored service this morning and everything has now caught up, but the outages meant that builds were processed quite differently from the usual order.  Since BSI just fetches the ten most recent build (ordered by the most recent change, so whichever of the started and finished times is newer) and files everything that's not the latest one for its architecture under "Previous builds", this is as expected.

It's possible that it would be worth expanding the batch of builds fetched by BSI from 10 to 15 now that we have 3 architectures rather than 2; I'll leave that up to more active committers and designers to consider.  Aside from that, I don't think there's much to be done about this except for possibly build sets.

**Comment by evandandrea** [comment]
Thanks Colin. @matthewpaulthomas, what do you think we should do here? I'm inclined to either suggest showing the five most recent builds for each architecture, or just showing all builds (behind a "show more" button).

**Comment by cjwatson** [comment]
The "five most recent builds for each architecture" approach is potentially tricky: if an architecture has only recently been added and only has (say) one build, then BSI has to go all the way to the end of the builds collection before it can tell that there are no more builds.  Any approach like that would require some different interfaces in Launchpad and some careful thought about performance (e.g. we probably don't want to have to make one webservice query per architecture).

Showing all builds doesn't have those complications, but it would require pagination (or a "show more" button, as long as that just retrieves the next batch rather than trying to show everything).  It may be worthwhile since it's otherwise somewhat hard to discover old builds.

**Comment by matthewpaulthomas** [comment]
“_n_ most recent builds for each architecture” would also be weird once (or now that? I can’t tell from the docs) you can specify build architectures in snapcraft.yaml. If you turned armel builds off six months ago, then turned them on again three revisions ago, a list of the five most recent builds for each architecture would include two ancient armel builds immediately below recent builds from other architectures.

I didn’t get around to specifying pagination for the repo page yet (neglecting my normal design checklist item, “What if this group of things contains 1000 items?”). But I imagined that the page would use lazy loading to show more builds in the table as you scrolled.

**Comment by nottrobin** [comment]
I don't know if this is still a problem in the new snapcraft.io/build experience. I'm moving it to snapcraft.io repo, it can be decided there.
